src/interpreter/mod.rs

```diff
@@ -172,6 +172,176 @@ impl<R: RealNumberInternalTrait> std::ops::Div<Number<R>> for Number<R> {
     }
 }
 
+impl<R: RealNumberInternalTrait> Number<R> {
+    pub fn sqrt(self) -> Self {
+        match self {
+            Number::Integer(num) => Number::Real(R::from(num).unwrap().sqrt()),
+            Number::Real(num) => Number::Real(num.sqrt()),
+            Number::Rational(a, b) => {
+                Number::Real(R::from(a).unwrap() / R::from(b).unwrap().sqrt())
+            }
+        }
+    }
+    pub fn floor(self) -> Self {
+        match self {
+            Number::Integer(num) => Number::Integer(num),
+            Number::Real(num) => Number::Real(num.floor()),
+            Number::Rational(a, b) => Number::Integer({
+                let quot = a / b;
+                if quot >= 0 || quot * b == a {
+                    quot
+                } else {
+                    quot - 1
+                }
+            }),
+        }
+    }
+
+    pub fn floor_quotient(self, rhs: Self) -> Result<Self> {
+        Ok((self / rhs)?.floor())
+    }
+
+    pub fn floor_remainder(self, rhs: Self) -> Result<Self> {
+        Ok(self - self.floor_quotient(rhs)? * rhs)
+    }
+
+    // Return an exact number that is numerically closest to the given number
+    pub fn exact(self) -> Result<Self> {
+        match self {
+            Number::Real(num) => match num.round().to_i32() {
+                Some(i) => Ok(Number::Integer(i)),
+                None => logic_error!("cannot be converted to an exact number"),
+            },
+            exact => Ok(exact),
+        }
+    }
+}
+#[test]
+fn number_floor() {
+    assert_eq!(Number::<f32>::Integer(5).floor(), Number::Integer(5));
+    assert_eq!(Number::<f32>::Rational(28, 3).floor(), Number::Integer(9));
+    assert_eq!(Number::<f32>::Rational(-43, 7).floor(), Number::Integer(-7));
+    assert_eq!(Number::<f32>::Rational(43, -7).floor(), Number::Integer(-7));
```

Comment:
Currently, Number::Rational can be (i32, i32), but from lexical, the can only be (i32, u32), it will be fix tn future, so do not test with negative divider.

Reply:
Fixed 8e720e2